Kudu.Services/Diagnostics/CrashDumpController.cs

```diff
@@ -0,0 +1,121 @@
+﻿using Kudu.Contracts.Diagnostics;
+using Kudu.Contracts.Settings;
+using Kudu.Contracts.Tracing;
+using Kudu.Core;
+using Kudu.Core.Infrastructure;
+using Kudu.Core.Tracing;
+using Kudu.Services.Infrastructure;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using System.Web.Http;
+
+namespace Kudu.Services.Diagnostics
+{
+    public class CrashDumpController : ApiController
+    {
+        private readonly ITracer _tracer;
+        private readonly IEnvironment _environment;
+        private readonly IDeploymentSettingsManager _settings;
+
+        public CrashDumpController(ITracer tracer,
+                                 IEnvironment environment,
+                                 IDeploymentSettingsManager settings)
+        {
+            _tracer = tracer;
+            _environment = environment;
+            _settings = settings;
+        }
+
+        [HttpGet]
+        public IEnumerable<CrashDumpInfo> List()
+        {
+            using (_tracer.Step("CrashDumpController.List"))
+            {
+                return FileSystemHelpers.GetFiles(_environment.CrashDumpsPath, "*.dmp")
+                    .Select(FileSystemHelpers.FileInfoFromFileName)
+                    .Select(f => GetCrashDumpInfoFromFile(f, Request.RequestUri.GetLeftPart(UriPartial.Authority)));
+            }
+        }
+
+        [HttpGet]
+        public HttpResponseMessage Get(string name)
+        {
+            using (_tracer.Step($"CrashDumpController.Get({name})"))
+            {
+                return Request.CreateResponse(HttpStatusCode.OK, GetCrashDump(name));
+            }
+        }
+
+        [HttpDelete]
+        public HttpResponseMessage Delete(string name)
+        {
+            using (_tracer.Step($"CrashDumpController.Delete({name})"))
+            {
+                var crashDump = GetCrashDump(name);
+                FileSystemHelpers.DeleteFile(crashDump.FilePath);
+                return Request.CreateResponse(HttpStatusCode.OK);
+            }
+        }
+
+        [HttpPost]
+        public async Task<HttpResponseMessage> Analyze(string name)
+        {
+            using (_tracer.Step($"CrashDumpController.Analyze({name})"))
+            {
+                var crashDump = GetCrashDump(name);
+                var exe = new Executable(ResolveCdbPath(), _environment.RootPath, _settings.GetCommandIdleTimeout());
+                using (MemoryStream outputStream = new MemoryStream(), errorStream = new MemoryStream())
+                {
+                    // ExecuteAsync could deadlock if called from an ASP.NET thread
+                    var exitCode = await Task.Run(async () => await exe.ExecuteAsync(_tracer, $"-c \"!analyze -v;q\" -z \"{crashDump.FilePath}\"", outputStream, errorStream));
```

Comment:
hmm I don't know. Here is the output from a console app that throws an exception and crashes. I can see the managed stack trace, but everything else is failing to load symbols.

```

Microsoft (R) Windows Debugger Version 6.3.9600.16384 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [D:\\home\\data\\dumps\\ConsoleApplication27.exe_151110_192049.dmp]
Comment: '
*** \"D:\\devtools\\sysinternals\\procdump.exe\" -accepteula -t 10088 -e -g 
*** Unhandled exception: E0434352.CLR'
User Mini Dump File: Only registers, stack and portions of memory are available

Symbol search path is: *** Invalid ***
****************************************************************************
* Symbol loading may be unreliable without a symbol search path.           *
* Use .symfix to have the debugger choose a symbol path.                   *
* After setting your symbol path, use .reload to refresh symbol locations. *
****************************************************************************
Executable search path is: 
Windows 8 Version 9200 UP Free x64
Product: Server, suite: TerminalServer DataCenter SingleUserTS
Built by: 6.2.9200.16384 (win8_rtm.120725-1247)
Machine Name:
Debug session time: Tue Nov 10 19:20:50.000 2015 (UTC + 0:00)
System Uptime: not available
Process Uptime: 0 days 0:00:16.000
..................................
This dump file has an exception of interest stored in it.
The stored exception information can be accessed via .ecxr.
(2768.2a8c): CLR exception - code e0434352 (first/second chance not available)
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for KERNELBASE.dll - 
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for clr.dll - 
KERNELBASE!RaiseException+0x68:
000007f9`73c1aea8 488b8c24c0000000 mov     rcx,qword ptr [rsp+0C0h] ss:00000095`99fbe750=00007d64ce7c3a9c
0:000> cdb: Reading initial command '.symfix;.reload;.loadby sos clr;~*k;!analyze -v;q'
..................................
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for KERNELBASE.dll - 
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for clr.dll - 

************* Symbol Loading Error Summary **************
Module name            Error
KERNELBASE             PDB not found : cache*
                       PDB not found : SRV*http://msdl.microsoft.com/download/symbols
clr                    PDB not found : cache*
                       PDB not found : SRV*http://msdl.microsoft.com/download/symbols

You can troubleshoot most symbol related issues by turning on symbol loading diagnostics (!sym noisy) and repeating the command that caused symbols to be loaded.
You should also verify that your symbol search path (.sympath) is correct.
----------------------------------------------------------------------------
The user dump currently examined is a minidump. Consequently, only a subset
of sos.dll functionality will be available. If needed, attaching to the live
process or debugging a full dump will allow access to sos.dll's full feature
set.
To create a full user dump use the command: .dump /ma <filename>
----------------------------------------------------------------------------

.  0  Id: 2768.2a8c Suspend: 0 Teb: 000007f5`ffb7e000 Unfrozen
Child-SP          RetAddr           Call Site
00000095`99fbe690 000007f9`6d1a2965 KERNELBASE!RaiseException+0x68
00000095`99fbe770 000007f9`6d1a276c clr!GetMetaDataPublicInterfaceFromInternal+0x15715
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for ntdll.dll - 
00000095`99fbe870 000007f9`0dac0531 clr!GetMetaDataPublicInterfaceFromInternal+0x1551c
00000095`99fbea20 000007f9`0dac04af 0x000007f9`0dac0531
00000095`99fbea60 000007f9`6d113883 0x000007f9`0dac04af
00000095`99fbeaa0 000007f9`6d113745 clr+0x3883
00000095`99fbeae0 000007f9`6d113cf6 clr+0x3745
00000095`99fbeb20 000007f9`6d22ed88 clr+0x3cf6
00000095`99fbec20 000007f9`6d22ef2a clr!GetCLRFunction+0xfd8
00000095`99fbee00 000007f9`6d22ee18 clr!GetCLRFunction+0x117a
00000095`99fbf0f0 000007f9`6d22efc0 clr!GetCLRFunction+0x1068
00000095`99fbf720 000007f9`6d22f086 clr!GetCLRFunction+0x1210
00000095`99fbf790 000007f9`6d241290 clr!GetCLRFunction+0x12d6
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for mscoreei.dll - 
00000095`99fbf820 000007f9`6e298389 clr!CorExeMain+0x14
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for mscoree.dll - 
00000095`99fbf860 000007f9`6e4d5b21 mscoreei!CorExeMain+0x5d
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for kernel32.dll - 
00000095`99fbf8b0 000007f9`74721842 mscoree!CorExeMain+0x69
00000095`99fbf8e0 000007f9`76bf0dfd kernel32!BaseThreadInitThunk+0x1a
00000095`99fbf910 00000000`00000000 ntdll!RtlUserThreadStart+0x21

   1  Id: 2768.2084 Suspend: 0 Teb: 000007f5`ffb7c000 Unfrozen
Child-SP          RetAddr           Call Site
00000095`9af4f7b8 000007f9`73bd129e ntdll!NtWaitForMultipleObjects+0xa
00000095`9af4f7c0 000007f9`6d23e47e KERNELBASE!GetTickCount+0x4e
00000095`9af4faa0 000007f9`6d23e3ac clr!InitializeFusion+0x60fe
00000095`9af4fb60 000007f9`6d23e26f clr!InitializeFusion+0x602c
00000095`9af4fbb0 000007f9`74721842 clr!InitializeFusion+0x5eef
00000095`9af4fc00 000007f9`76bf0dfd kernel32!BaseThreadInitThunk+0x1a
00000095`9af4fc30 00000000`00000000 ntdll!RtlUserThreadStart+0x21

   2  Id: 2768.2bf8 Suspend: 0 Teb: 000007f5`ffb7a000 Unfrozen
Child-SP          RetAddr           Call Site
00000095`b36bf6b8 000007f9`73bd129e ntdll!NtWaitForMultipleObjects+0xa
00000095`b36bf6c0 000007f9`6d224d8a KERNELBASE!GetTickCount+0x4e
00000095`b36bf9a0 000007f9`6d224ec9 clr!GetPrivateContextsPerfCounters+0x306da
00000095`b36bf9e0 000007f9`6d114299 clr!GetPrivateContextsPerfCounters+0x30819
00000095`b36bfa20 000007f9`6d11418c clr+0x4299
00000095`b36bfa60 000007f9`6d1140be clr+0x418c
00000095`b36bfb60 000007f9`6d22f237 clr+0x40be
00000095`b36bfbf0 000007f9`6d20b626 clr!GetCLRFunction+0x1487
00000095`b36bfc90 000007f9`74721842 clr!GetPrivateContextsPerfCounters+0x16f76
00000095`b36bfd50 000007f9`76bf0dfd kernel32!BaseThreadInitThunk+0x1a
00000095`b36bfd80 00000000`00000000 ntdll!RtlUserThreadStart+0x21

   3  Id: 2768.29c8 Suspend: 0 Teb: 000007f5`ffb78000 Unfrozen
Child-SP          RetAddr           Call Site
00000095`b3d9f6b8 000007f9`76bb76c3 ntdll!ZwWaitForWorkViaWorkerFactory+0xa
00000095`b3d9f6c0 000007f9`74721842 ntdll!RtlTryEnterCriticalSection+0x833
00000095`b3d9f960 000007f9`76bf0dfd kernel32!BaseThreadInitThunk+0x1a
00000095`b3d9f990 00000000`00000000 ntdll!RtlUserThreadStart+0x21
*******************************************************************************
*                                                                             *
*                        Exception Analysis                                   *
*                                                                             *
*******************************************************************************

***** OS symbols are WRONG. Please fix symbols to do analysis.

*************************************************************************
***                                                                   ***
***                                                                   ***
***    Either you specified an unqualified symbol, or your debugger   ***
***    doesn't have full symbol information.  Unqualified symbol      ***
***    resolution is turned off by default. Please either specify a   ***
***    fully qualified symbol module!symbolname, or enable resolution ***
***    of unqualified symbols by typing \".symopt- 100\". Note that   ***
***    enabling unqualified symbol resolution with network symbol     ***
***    server shares in the symbol path may cause the debugger to     ***
***    appear to hang for long periods of time when an incorrect      ***
***    symbol name is typed or the network symbol server is down.     ***
***                                                                   ***
***    For some commands to work properly, your symbol path           ***
***    must point to .pdb files that have full type information.      ***
***                                                                   ***
***    Certain .pdb files (such as the public OS symbols) do not      ***
***    contain the required information.  Contact the group that      ***
***    provided you with these symbols if you need this command to    ***
***    work.                                                          ***
***                                                                   ***
***    Type referenced: ntdll!_PEB                                    ***
***                                                                   ***
*************************************************************************
*** WARNING: Unable to verify checksum for mscorlib.ni.dll
*** ERROR: Module load completed but symbols could not be loaded for mscorlib.ni.dll

************* Symbol Loading Error Summary **************
Module name            Error
mscorlib.ni            PDB not found : cache*
                       PDB not found : SRV*http://msdl.microsoft.com/download/symbols

You can troubleshoot most symbol related issues by turning on symbol loading diagnostics (!sym noisy) and repeating the command that caused symbols to be loaded.
You should also verify that your symbol search path (.sympath) is correct.
*************************************************************************
***                                                                   ***
***                                                                   ***
***    Either you specified an unqualified symbol, or your debugger   ***
***    doesn't have full symbol information.  Unqualified symbol      ***
***    resolution is turned off by default. Please either specify a   ***
***    fully qualified symbol module!symbolname, or enable resolution ***
***    of unqualified symbols by typing \".symopt- 100\". Note that   ***
***    enabling unqualified symbol resolution with network symbol     ***
***    server shares in the symbol path may cause the debugger to     ***
***    appear to hang for long periods of time when an incorrect      ***
***    symbol name is typed or the network symbol server is down.     ***
***                                                                   ***
***    For some commands to work properly, your symbol path           ***
***    must point to .pdb files that have full type information.      ***
***                                                                   ***
***    Certain .pdb files (such as the public OS symbols) do not      ***
***    contain the required information.  Contact the group that      ***
***    provided you with these symbols if you need this command to    ***
***    work.                                                          ***
***                                                                   ***
***    Type referenced: nt!IMAGE_NT_HEADERS32                         ***
***                                                                   ***
*************************************************************************
*** WARNING: Unable to verify checksum for ConsoleApplication27.exe
*** ERROR: Module load completed but symbols could not be loaded for ConsoleApplication27.exe
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for ole32.dll - 

************* Symbol Loading Error Summary **************
Module name            Error
ole32                  PDB not found : cache*
                       PDB not found : SRV*http://msdl.microsoft.com/download/symbols

You can troubleshoot most symbol related issues by turning on symbol loading diagnostics (!sym noisy) and repeating the command that caused symbols to be loaded.
You should also verify that your symbol search path (.sympath) is correct.
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for combase.dll - 

************* Symbol Loading Error Summary **************
Module name            Error
combase                PDB not found : cache*
                       PDB not found : SRV*http://msdl.microsoft.com/download/symbols

You can troubleshoot most symbol related issues by turning on symbol loading diagnostics (!sym noisy) and repeating the command that caused symbols to be loaded.
You should also verify that your symbol search path (.sympath) is correct.

FAULTING_IP: 
KERNELBASE!RaiseException+68
000007f9`73c1aea8 488b8c24c0000000 mov     rcx,qword ptr [rsp+0C0h]

EXCEPTION_RECORD:  ffffffffffffffff -- (.exr 0xffffffffffffffff)
ExceptionAddress: 000007f973c1aea8 (KERNELBASE!RaiseException+0x0000000000000068)
   ExceptionCode: e0434352 (CLR exception)
  ExceptionFlags: 00000001
NumberParameters: 5
   Parameter[0]: ffffffff80131500
   Parameter[1]: 0000000000000000
   Parameter[2]: 0000000000000000
   Parameter[3]: 0000000000000000
   Parameter[4]: 000007f96d110000

CONTEXT:  0000000000000000 -- (.cxr 0x0;r)
rax=0000000000000009 rbx=0000000000000001 rcx=000000959a5e0000
rdx=0000000000000090 rsi=0000009599fbe808 rdi=0000000000000005
rip=000007f973c1aea8 rsp=0000009599fbe690 rbp=0000009599fbeac0
 r8=000007f900000000  r9=fffffffffffffffe r10=000007f96d2a6379
r11=ffffffff00000002 r12=0000000000004000 r13=0000009599fbeb40
r14=00000000e0434352 r15=0000000000000004
iopl=0         nv up ei pl nz na pe nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000200
KERNELBASE!RaiseException+0x68:
000007f9`73c1aea8 488b8c24c0000000 mov     rcx,qword ptr [rsp+0C0h] ss:00000095`99fbe750=00007d64ce7c3a9c

DEFAULT_BUCKET_ID:  WRONG_SYMBOLS

PROCESS_NAME:  ConsoleApplication27.exe

ADDITIONAL_DEBUG_TEXT:  \r
You can run '.symfix; .reload' to try to fix the symbol path and load symbols.

FAULTING_MODULE: 000007f976bb0000 ntdll

DEBUG_FLR_IMAGE_TIMESTAMP:  0

ERROR_CODE: (NTSTATUS) 0xe0434352 - <Unable to get error code text>

EXCEPTION_CODE: (NTSTATUS) 0xe0434352 - <Unable to get error code text>

EXCEPTION_PARAMETER1:  ffffffff80131500

EXCEPTION_PARAMETER2:  0000000000000000

EXCEPTION_PARAMETER3:  0000000000000000

EXCEPTION_PARAMETER4: 0

APP:  consoleapplication27.exe

ANALYSIS_VERSION: 6.3.9600.16384 (debuggers(dbg).130821-1623) amd64fre

MANAGED_STACK: 
    SP               IP               Function
    0000009599FBEA20 000007F90DAC0531 ConsoleApplication27!ConsoleApplication27.Program.Throw()+0x61
    0000009599FBEA60 000007F90DAC04AF ConsoleApplication27!ConsoleApplication27.Program.Main(System.String[])+0x2f

MANAGED_STACK_COMMAND:  !do 0


LAST_CONTROL_TRANSFER:  from 000007f96d1a2965 to 000007f973c1aea8

PRIMARY_PROBLEM_CLASS:  WRONG_SYMBOLS

BUGCHECK_STR:  APPLICATION_FAULT_WRONG_SYMBOLS_CLR_EXCEPTION

STACK_TEXT:  
00000095`99fbea20 000007f9`0dac0531 consoleapplication27!ConsoleApplication27.Program.Throw+0x61
00000095`99fbea60 000007f9`0dac04af consoleapplication27!ConsoleApplication27.Program.Main+0x2f


FOLLOWUP_IP: 
+68
000007f9`0dac04af 90              nop

SYMBOL_STACK_INDEX:  1

SYMBOL_NAME:  consoleapplication27!ConsoleApplication27.Program.Main+2f

FOLLOWUP_NAME:  MachineOwner

MODULE_NAME: consoleapplication27

IMAGE_NAME:  ConsoleApplication27.exe

STACK_COMMAND:  !do 0
 ; ** Pseudo Context ** ; kb

BUCKET_ID:  WRONG_SYMBOLS

FAILURE_BUCKET_ID:  WRONG_SYMBOLS_e0434352_ConsoleApplication27.exe!ConsoleApplication27.Program.Main

ANALYSIS_SOURCE:  UM

FAILURE_ID_HASH_STRING:  um:wrong_symbols_e0434352_consoleapplication27.exe!consoleapplication27.program.main

FAILURE_ID_HASH:  {9aaa794a-0d8c-c3f9-3096-2a1e227c4661}

Followup: MachineOwner
---------

quit:

```